tinygrad/codegen/kernel.py

```diff
@@ -607,97 +607,82 @@ def name(self) -> str:
     return name + colored(num, 'BLACK')
 
   def get_optimized_ast(self) -> UOp:
-    # set the shapetrackers to the optimized ones, fixup reduceop
-    # transformed to the final UOp
     @functools.lru_cache(None)
-    def fixup_ast(op:UOp, apply_to_st=None) -> UOp:
-      arg = op.arg
-      if op.op in GroupOp.Buffer:
-        # for locals, we use the ShapeTracker that's in the srcs
-        st = op.st_arg if op.src[0].op is Ops.DEFINE_LOCAL else self.sts[self.bufs.index(op)]
-        st_uop = (st if apply_to_st is None else apply_to_st(st)).to_uop()
-        if op.op is Ops.VALID: return op.replace(src=(st_uop,))
-        if op.op is Ops.STORE: return op.replace(src=(op.src[0], st_uop, fixup_ast(op.src[2], apply_to_st)))
-        return op.replace(src=(op.src[0], st_uop, *[fixup_ast(x, apply_to_st) for x in op.src[2:]]))
+    def fixup_ast(op:UOp) -> UOp:
+      ret = op.replace(src=tuple(fixup_ast(x) for x in op.src))
+      if op.op in GroupOp.Buffer and op in self.bufs: return ret.view(self.sts[self.bufs.index(op)])
```

Comment:
what's "and op in self.bufs"

Reply:
that's for bufs who's st have been modified within kernel (through opt opts). Before it checked the source was not DEFINE_LOCAL but found it cleaner this way.

Reply:
can you narrow it down to that specific local case? this can hide a legit bug.

Reply:
before: `st = op.st_arg if op.src[0].op is Ops.DEFINE_LOCAL else self.sts[self.bufs.index(op)]`
after: 
if the buf is in op then there is a st associated to it on self.sts

Reply:
> can you narrow it down to that specific local case? this can hide a legit bug.

For DEFINE_LOCAL no fixing the st is needed. There is no special case. At least in this stage of st fixing. Maybe I'm missing something though